src/object_store/cache.rs

```diff
@@ -217,88 +223,201 @@ impl CachingObjectStore {
+                                })?;
+                                debug!("{key:?} awaited");
+                                rx.borrow().clone().map_err(|e| {
+                                    object_store::Error::Generic {
+                                        store: "cache_store",
+                                        source: Box::new(e),
+                                    }
+                                })?
+                            }
+                            Err(_) => {
+                                debug!("{key:?} refetching");
+                                // TODO: maybe check that value is not in cache now,
+                                // and if not insert + trigger persisting to file
+                                let data = self
+                                    .inner
+                                    .get_range(path, chunk_range.clone())
+                                    .await?;
+                                debug!("{key:?} refetched");
+                                data
+                            }
                         }
                     }
+                    CacheValue::Memory(data) => {
+                        debug!("Cache value for {key:?} fetched from memory");
+                        data.clone()
+                    }
+                    CacheValue::File(path, _) => {
+                        debug!("Cache value for {key:?} fetched from file");
+                        self.file_manager.read_file(path).await.unwrap()
+                    }
                 });
+            } else {
+                // Entry is fresh; we need to evaluate it in the upcoming batch...
+                chunk_batches.push((key.clone(), tx));
+                // ... and also create a new sender/receiver for the next pending entry
+                (tx, rx) = channel(Ok(Bytes::new()));
+            }
 
-                // While the write task runs cache the in-memory bytes and serve that to all calls
-                // prior to the write task completing.
-                debug!("Caching value for ({path:?}, {chunk}) in memory");
-                Ok(CacheValue::Memory(data))
-            })
-            .await
-            .map_err(|e| object_store::Error::Generic {
-                store: "cache_store",
-                source: Box::new(e),
-            })?;
-
-        match value {
-            CacheValue::File(path, _) => {
-                debug!("Cache value for ({path:?}, {chunk}) fetched from the file");
-                let data = self.file_manager.read_file(path).await.unwrap();
-                Ok(data)
+            if !chunk_batches.is_empty() && (!is_fresh || chunk == end_chunk) {
+                // Evaluate the accumulated chunk batch if either the current entry is not fresh
+                // (meaning we can not extend the batch anymore and keep it continuous), or
+                // we've reached the last chunk (meaning there are not more chunks to be batched).
```

Comment:
As discussed, possible that DF already does this pre-coalescing for us (https://github.com/apache/arrow-rs/blob/a999fb86764e9310bb4822c7e7c6551f247e0e0b/object_store/src/lib.rs#L573), but we'll need to look at some read stats to figure out what we need to do besides just providing a cache